### PR TITLE
fix lexrange issue, added lexrange to TAGS, added the ability to specify includeMin and includeMax on lexrange

### PR DESCRIPTION
**Pull request by MeirShpilraien**

### CODE REVIEW

File: src/cpptests/t_llapi.cpp [hunk 1]
```diff
@@ -286,7 +286,7 @@ TEST_F(LLApiTest, testRanges) {
     RediSearch_SpecAddDocument(index, d);
   }
 
-  RSQNode* qn = RediSearch_CreateLexRangeNode(index, FIELD_NAME_1, "MarkN", "MarkX");
+  RSQNode* qn = RediSearch_CreateLexRangeNode(index, FIELD_NAME_1, "MarkN", "MarkX", 1, 1);
```
Comment (mnunberg) [inline src/cpptests/t_llapi.cpp]: I'd prefer flags, e.g. `REDISEARCH_RANGE_EXCLUDE_MIN` and `REDISEARCH_RANGE_INCLUDE_MAX`. Default should still be as before, where the begin is inclusive but the end is exclusive.

File: src/cpptests/t_llapi.cpp [hunk 2]
```diff
@@ -298,8 +298,88 @@ TEST_F(LLApiTest, testRanges) {
...
+    sprintf(namebuf, "doc%c", c);
+    ASSERT_NE(results.end(), results.find(namebuf));
+  }
+  RediSearch_ResultsIteratorFree(iter);
+
+  // printf("Have %lu ids in range!\n", results.size());
+  RediSearch_DropIndex(index);
+}
+
+TEST_F(LLApiTest, testRangesOnTags) {
```
Comment (mnunberg) [inline src/cpptests/t_llapi.cpp]: I'm sure there's a way we can refactor out some common code between those two test functions.

File: src/cpptests/t_trie.cpp [hunk 3]
```diff
@@ -24,11 +24,24 @@ static ElemSet trieIterRange(Trie *t, const char *begin, size_t nbegin, const ch
...
+  }
+
+  if (!end) {
+    r2Ptr = NULL;
+    nr2 = -1;
+  }
+
   ElemSet foundElements;
-  TrieNode_IterateRange(t->root, r1, nr1, r2, nr2,
+  TrieNode_IterateRange(t->root, r1Ptr, nr1, true, r2Ptr, nr2, false,
```
Comment (mnunberg) [inline src/cpptests/t_trie.cpp]: again, use flags

File: src/cpptests/t_trie.cpp [hunk 4]
```diff
@@ -66,10 +79,10 @@ TEST_F(TrieTest, testBasicRange) {
 
   // Min and max the same- should not yield anything
```
Comment (mnunberg) [inline src/cpptests/t_trie.cpp]: I guess this comment is no longer relevant

File: src/query.c [hunk 5]
```diff
@@ -348,8 +348,29 @@ typedef struct {
   size_t cap;
   QueryEvalCtx *q;
   QueryNodeOptions *opts;
+  double weight;
 } LexRangeCtx;
 
+static void rangeIterCbStrs(const char *r, size_t n, void *p, void *invidx) {
```
Comment (mnunberg) [inline src/query.c]: Refactor out common parts; I can't tell the difference between the two functions.

File: src/query.c [hunk 6]
```diff
@@ -838,7 +889,7 @@ void QueryNode_AddChildren(QueryNode *n, QueryNode **children, size_t nchildren)
   if (n->type == QN_TAG) {
     for (size_t ii = 0; ii < nchildren; ++ii) {
       if (children[ii]->type == QN_TOKEN || children[ii]->type == QN_PHRASE ||
-          children[ii]->type == QN_PREFX) {
+          children[ii]->type == QN_PREFX || children[ii]->type == QN_LEXRANGE) {
```
Comment (mnunberg) [inline src/query.c]: good catch

File: src/trie/trie.c [hunk 7]
```diff
@@ -669,57 +668,60 @@ typedef struct {
...
+  r->offset += n->len;
+  r->buf[r->offset] = 0;
+
+  if (__trieNode_isTerminal(n)) {
+    r->callback(r->buf, r->offset, r->cbctx);
+  }
+
+  TrieNode **arr = __trieNode_children(n);
+
+  for (int ii = 0; ii < n->numChildren; ++ii) {
```
Comment (mnunberg) [inline src/trie/trie.c]: size_t

File: src/util/bsearch.h [hunk 8]
```diff
@@ -35,8 +35,10 @@ typedef int (*rsbcompare)(const void *s, const void *elem);
  * @param cmp the comparison function
  * @return `end`
  */
-static inline size_t rsb_ge(const void *arr, size_t narr, size_t elemsz, size_t begin, size_t end,
-                            const void *s, rsbcompare cmp) {
+static inline int rsb_gr(const void *arr, size_t narr, size_t elemsz, const void *s,
```
Comment (mnunberg) [inline src/util/bsearch.h]: why the name change?

File: src/cpptests/t_llapi.cpp (hunk 1 above)
Comment (MeirShpilraien) [reply src/cpptests/t_llapi.cpp]: This is how the LEXNODE is defines I want to keep them the same..

File: src/cpptests/t_trie.cpp (hunk 3 above)
Comment (MeirShpilraien) [reply src/cpptests/t_trie.cpp]: Same comment as above

File: src/util/bsearch.h (hunk 8 above)
Comment (MeirShpilraien) [reply src/util/bsearch.h]: Now it check for Greater only and not GE

**Comment by mnunberg** [comment]
use `gt`, i.e. "Greater Than"

**Pull request merged**